Fix setting bounds (non-canonical) untagged capabilities

arichardson opened the pull request:
This was found by the fuzzer since it triggered an assertion, now it
matches the sail outputs.
When calling decompress_raw we have to use req_base and not the current
address to infer the bounds. This did not matter previously since we only
called the function on in-bounds caps, so the result was the same, but now
that we allow out-of-bounds untagged values, we have to use req_base
instead and set the address to req_base.

arichardson committed:
Extend the fuzzer to also compare setting bounds against sail

arichardson committed:
Restrict _cc_N(setbounds_impl) validity assertions to tagged capabilities

When called on untagged quantities these do not necessarily hold.

arichardson committed:
Fix and add test for setting bounds on a non-canonical untagged capability

This was found by the fuzzer since it triggered an assertion, now it
matches the sail outputs.
When calling decompress_raw we have to use req_base and not the current
address to infer the bounds. This did not matter previously since we only
called the function on in-bounds caps, so the result was the same, but now
that we allow out-of-bounds untagged values, we have to use req_base
instead and set the address to req_base.

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -897,6 +900,7 @@ static inline bool _cc_N(setbounds_impl)(_cc_cap_t* cap, _cc_addr_t req_base, _c
 
 /* @return whether the operation was able to set precise bounds precise or not */
 static inline bool _cc_N(setbounds)(_cc_cap_t* cap, _cc_addr_t req_base, _cc_length_t req_top) {
+    cap->_cr_cursor = req_base;
```
This is odd

arichardson replied:
To match https://github.com/CTSRD-CHERI/sail-cheri-riscv/blob/41454cad0ef41ec8e367be22f721517818224fc6/src/cheri_cap_common.sail#L368, but I guess I could rewrite this to move the assignment to _impl.

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -829,24 +829,27 @@ static inline bool _cc_N(setbounds_impl)(_cc_cap_t* cap, _cc_addr_t req_base, _c
-           "Must be used on inbounds caps or request zero-length cap at top");
-    assert((cursor >= cap->cr_base) && "Must be used on inbounds caps");
+    if (cap->cr_tag) {
+        const _cc_addr_t cursor = cap->_cr_cursor;
```
This will be unused for NDEBUG

jrtc27 commented on cheri_compressed_cap_common.h:
```diff
@@ -829,24 +829,27 @@ static inline bool _cc_N(setbounds_impl)(_cc_cap_t* cap, _cc_addr_t req_base, _c
-    assert((cursor >= cap->cr_base) && "Must be used on inbounds caps");
+    if (cap->cr_tag) {
+        const _cc_addr_t cursor = cap->_cr_cursor;
+        _cc_debug_assert(req_base == cap->_cr_cursor && "Must set address to req_base when calling this function");
```
This seems a little odd given we set _cr_cursor to req_base later

arichardson commented:
Rebased and fixed issues.

arichardson committed:
Relax another assertion for untagged inputs

Due to the way we decode top/bot it is possible for top to decode to
1<<66, but that is then truncated to zero and can therefore be less than
the base of the capability. The example test case was found by TestRIG

See https://github.com/CTSRD-CHERI/sail-cheri-riscv/pull/36 for an
alternative decoding approach that guarantees top >= base.

arichardson committed:
Improve setbounds fuzzer coverage

Actually request more than 2 addresses worth of input, so we end up with
a non-zero req_len input. With this change the assertion relaxed in the
last commit is also found by libfuzzer.

arichardson commented:
Found another assertion that can be triggered on untagged inputs. Relaxing this also seems fine unless we decide to merge https://github.com/CTSRD-CHERI/sail-cheri-riscv/pull/36.

arichardson commented:
If there are no further comments I will merge this tomorrow.
